📚 Update migration guide (#5420)
apollographql/apollo-kotlin: docs/source/migration/4.0.mdx

We recommend using the plugin to automate the repetitive tasks but still go through this document for the details. 

## Error handling

> ⚠️ Error handling changes are a behaviour change that is not be detected at compile time. Usages of `execute`, `toFlow` and `watch` must be updated to the new error handling or changed to their v3 compat equivalent.

### Fetch errors do not throw

In Apollo Kotlin 3.x, non-GraphQL errors like network errors, cache misses, and parsing errors were surfaced by throwing exceptions in `ApolloCall.execute()` and in Flows (`ApolloCall.toFlow()`, `ApolloCall.watch()`). This was problematic because it was a difference in how to handle GraphQL errors vs other errors. Moreover, throwing terminates a Flow and consumers would have to handle re-collection.

In Apollo Kotlin 4.0, a new field `ApolloResponse.exception` has been added and these errors are now surfaced by returning (for `execute()`) or emitting (for Flows) an `ApolloResponse` with a non-null `exception` instead of throwing it. 

This allows consumers to handle different kinds of errors at the same place, and it prevents Flows from being terminated. 

> Queries and mutations: 

Author: 👍 good catch

```kotlin
// schema/build.gradle.kts

// Replace
apollo {
  service("service") {
    // ...
    
    // Generate all the types in the schema module
    alwaysGenerateTypesMatching.set(listOf(".*"))

    // Enable generation of metadata for use by downstream modules 
    generateApolloMetadata.set(true)
  }
}

// With
apollo {
  service("service") {
    // ...

    // Enable generation of metadata for use by downstream modules 
    generateApolloMetadata.set(true)

    // Get used types from the downstream module1
    isADependencyOf(project(":feature1"))

    // Get used types from the downstream module2
    isADependencyOf(project(":feature2"))
    
    // ...
  }
}
```

If you were using `apolloUsedCoordinates`, you can also remove it:

```kotlin
dependencies {
  // Remove this 
  apolloUsedCoordinates(project(":feature1"))
}
```

### Custom scalars declaration

> `customScalarsMapping` is removed and replaced with `mapScalar()` that allows mapping to built-in types easier and/or provide a compile type adapter for a given type:

Reviewer: ```suggestion
`customScalarsMapping` is removed and replaced with `mapScalar()` which makes it easier to map to built-in types and/or provide a compile type adapter for a given type:
```
